dbms/src/DataStreams/WindowTransformAction.cpp

```diff
@@ -1256,17 +1293,79 @@ void WindowTransformAction::appendBlock(Block & current_block)
     auto & window_block = window_blocks.back();
     window_block.rows = current_block.rows();
 
-    // Initialize output columns and add new columns to output block.
-    for (auto & ws : workspaces)
+    for (const auto & return_type : return_types)
     {
-        MutableColumnPtr res = ws.window_function->getReturnType()->createColumn();
+        MutableColumnPtr res = return_type->createColumn();
         res->reserve(window_block.rows);
         window_block.output_columns.push_back(std::move(res));
     }
 
     window_block.input_columns = current_block.getColumns();
 }
 
+bool WindowTransformAction::checkIfNeedDecrease()
+{
+    if (first_processed)
+        return false;
+
+    if (prev_frame_end <= frame_start)
+        return false;
+
+    // added row refers to the rows that have been added in the previous frame
+    UInt64 add_row_num = distance(prev_frame_end, frame_start);
+    UInt64 decrease_row_num = distance(frame_start, prev_frame_start);
+    return add_row_num > decrease_row_num;
```

Comment:
I don't understand why `add_row_num > decrease_row_num` means we need to decrease. 
As I understand it, `add_row_num` represents the number of rows from the previous frame that are still included in the current frame, while `decrease_row_num` represents how many rows the `frame_start` has moved forward. 
Why does comparing these two values determine whether we need to decrease?

Reply:
can you add more comments?

Reply:
> I don't understand why `add_row_num > decrease_row_num` means we need to decrease. As I understand it, `add_row_num` represents the number of rows from the previous frame that are still included in the current frame, while `decrease_row_num` represents how many rows the `frame_start` has moved forward. Why does comparing these two values determine whether we need to decrease?

Suppose add_row_num is 10 and decrease_row_num is 5, we need to add 10 times and if we choose to decrease, we will only need to decrease for 5 times. We always choose the operation that need less steps.

Reply:
> can you add more comments?

okk